typing/typedecl.ml

```diff
@@ -1896,15 +1902,21 @@ let update_decl_jkind env dpath decl =
     | Type_variant (cstrs, rep) ->
       let cstrs, rep, type_jkind = update_variant_kind cstrs rep in
       let type_jkind, type_has_illegal_crossings = add_crossings type_jkind in
+      let type_jkind = Jkind.mark_best type_jkind in
       { decl with type_kind = Type_variant (cstrs, rep);
                   type_jkind;
                   type_has_illegal_crossings },
       type_jkind
   in
 
+
   (* check that the jkind computed from the kind matches the jkind
      annotation, which was stored in decl.type_jkind *)
   if new_jkind != decl.type_jkind then begin
+    (* Add the decl to a temporary env, so that normalization can look up the kind of the
+       type itself. *)
+    let env = add_type ~check:true ~shape id new_decl env in
```

Comment:
This seems pretty fragile, in that it will work only for simple recursion, not mutual recursion. Is this important for correctness? If so, then I fear we're not correct (because of the failure to account for mutual recursion). Or is this just an optimization to improve normalization? Then there should be a comment explaining this.